File: src/main/java/com/learn/todolist/application/TodoApplication.java
package com.learn.todolist.application;

import com.learn.todolist.api.dto.TodoCreateReq;
import com.learn.todolist.api.dto.TodoModifyReq;
import com.learn.todolist.domain.Todo;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collection;
import java.util.UUID;

@RequiredArgsConstructor
@Service
public class TodoApplication {

    private final TodoManagementServiceImpl todoManagementService;

    @Transactional
    public CommonResponse processCreateTodo(TodoCreateReq request) {
        UUID userId = UUID.fromString(SecurityContextHolder.getContext().getAuthentication().getPrincipal().toString());
        boolean result= todoManagementService.createTodo(userId, request.title(), request.content());
        return new CommonResponse("Todo Created", result);
    }

    @Transactional
    public CommonResponse processUpdateTodo(Long id, TodoModifyReq request) {
        boolean result = todoManagementService.updateTodo(id, request.title(), request.content(), request.status());
        return new CommonResponse("Todo Updated", result);
    }

    @Transactional
    public CommonResponse processDeleteTodo(Long id) {
        boolean result = todoManagementService.deleteTodo(id);
        return new CommonResponse("Todo Deleted", result);
    }

    public CommonResponse processGetTodoDetail(Long id) {
        Todo todo = todoManagementService.getTodoDetail(id);
        return new CommonResponse("Todo Detail", todo);
    }

    public CommonResponse processGetTodoList() {
        UUID userId = UUID.fromString(SecurityContextHolder.getContext().getAuthentication().getPrincipal().toString());
        Collection<Todo> todoList = todoManagementService.getUserTodoList(userId);
        return new CommonResponse("Todo List", todoList);
    }
}

File: src/main/kotlin/com/todolist/aplication/todo/usecase/TodoApplication.kt
package com.todolist.aplication.todo.usecase

import com.todolist.aplication.todo.dto.request.TodoCreateRequest
import com.todolist.aplication.todo.dto.request.TodoUpdateRequest
import com.todolist.domain.common.CommonResponse
import com.todolist.domain.todo.module.TodoService
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

private val logger = KotlinLogging.logger {}

@Service
class TodoApplication(
    private val todoService: TodoService
) {
    @Transactional
    fun processCreateTodo(request: TodoCreateRequest): CommonResponse {
        logger.info { "Create Todo Start" }
        val todo = todoService.createTodo(title = request.title, content = request.content)
        logger.info { "Create Todo Done" }
        return CommonResponse(
            message = "할일 생성 성공",
            result = todo
        )
    }

    @Transactional
    fun processUpdateTodo(request: TodoUpdateRequest): CommonResponse {
        logger.info { "Update Todo Start" }
        val todo = todoService.updateTodo(id = request.id, content = request.content, newStatus = request.status)
        logger.info { "Update Todo Done" }
        return CommonResponse(
            message = "할일 수정 성공",
            result = todo
        )
    }

    @Transactional
    fun processDeleteTodo(id: String): CommonResponse {
        logger.info { "Delete Todo Start" }
        val todo = todoService.deleteTodo(id=id)
        logger.info { "Delete Todo Done" }
        return CommonResponse(
            message = "할일 삭제 성공",
            result = todo
        )
    }

    fun processGetRecentTodo(): CommonResponse {
        logger.info { "Get Recent Todo Start" }
        val todo = todoService.getMostRecentTodo()
        logger.info { "Get Recent Todo Done" }
        return CommonResponse(
            message = "가장 최근 할일 조회 성공",
            result = todo
        )
    }

    fun processGetTodoDetail(id: String): CommonResponse {
        logger.info { "Get Todo Detail Start" }
        val todo = todoService.getTodoDetail(id = id)
        logger.info { "Get Todo Detail Done" }
        return CommonResponse(
            message = "할일 상세 조회 성공",
            result = todo
        )
    }

    fun processGetAllTodo(): CommonResponse {
        logger.info { "Get All Todo Start" }
        val todo = todoService.getAllTodo()
        logger.info { "Get All Todo Done" }
        return CommonResponse(
            message = "할일 전체 조회 성공",
            result = todo
        )

    }
}